src/main/java/de/zalando/aruha/nakadi/controller/EventStreamController.java

```diff
@@ -0,0 +1,162 @@
+            @RequestParam(value = "batch_flush_timeout", required = false, defaultValue = "30") final int batchTimeout,
+            @RequestParam(value = "stream_timeout", required = false, defaultValue = "0") final int streamTimeout,
+            @RequestParam(value = "stream_keep_alive_limit", required = false, defaultValue = "0") final int streamKeepAliveLimit,
+            @Nullable @RequestHeader(name = "X-Flow-Id", required = false) final String flowId,
+            @Nullable @RequestHeader(name = "X-nakadi-cursors", required = false) final String cursorsStr,
+            final HttpServletRequest request, final HttpServletResponse response) throws IOException {
+
+        LOG.trace("starting event stream for flow id: {}", flowId);
+
+        return outputStream -> {
+            try {
+                // todo: we should get topic from EventType after persistence of EventType is implemented
+                @SuppressWarnings("UnnecessaryLocalVariable")
+                final String topic = eventTypeName;
+
+                // validate parameters
+                if (!topicRepository.topicExists(topic)) {
+                    writeProblemResponse(response, outputStream, HttpStatus.NOT_FOUND.value(),
+                            new Problem("topic not found"));
+                    return;
+                }
+                if (streamLimit != 0 && streamLimit < batchLimit) {
+                    writeProblemResponse(response, outputStream, HttpStatus.UNPROCESSABLE_ENTITY.value(),
+                            new Problem("stream_limit can't be lower than batch_limit"));
+                    return;
+                }
+                if (streamTimeout != 0 && streamTimeout < batchTimeout) {
+                    writeProblemResponse(response, outputStream, HttpStatus.UNPROCESSABLE_ENTITY.value(),
+                            new Problem("stream_timeout can't be lower than batch_flush_timeout"));
+                    return;
+                }
+
+                // deserialize cursors
+                Optional<List<Cursor>> cursors = Optional.empty();
+                if (cursorsStr != null) {
+                    try {
+                        cursors = Optional.of(jsonMapper.<List<Cursor>>readValue(cursorsStr,
+                                new TypeReference<ArrayList<Cursor>>() {}));
+                    } catch (IOException e) {
+                        writeProblemResponse(response, outputStream, HttpStatus.BAD_REQUEST.value(),
+                                new Problem("incorrect syntax of X-nakadi-cursors header"));
+                        return;
+                    }
+                }
+
+                // check that offsets are not out of bounds
+                if (cursors.isPresent() && !topicRepository.areCursorsCorrect(topic, cursors.get())) {
+                    writeProblemResponse(response, outputStream, HttpStatus.UNPROCESSABLE_ENTITY.value(),
```

Comment:
Unprocessable entity would be more of a semantic issue (one can't make sense of the cursor, like missing data); looking up, I believe is more correct to say 412 Precondition Failed:

>    The 412 (Precondition Failed) status code indicates that one or more
>    conditions given in the request header fields evaluated to false when
>    tested on the server.  This response code allows the client to place
>    preconditions on the current resource state (its current
>    representations and metadata) and, thus, prevent the request method
>    from being applied if the target resource is in an unexpected state.

https://tools.ietf.org/html/rfc7232#section-4.2
Will include this information on the API too.